app/services/hashtag_autocomplete_service.rb

```diff
@@ -206,24 +219,20 @@ def search(
 
     return limited_results if limited_results.length >= limit
 
-    # Search the data source for each type, validate and sort results,
-    # and break off from searching more data sources if we reach our limit
+    # Next priority are slugs which start with the search term.
```

Comment:
I am wondering if we could leverage SQL to execute fewer queries instead of having one query per object per type of search.

```sql
SELECT ...
       CASE WHEN name = 'TERM' THEN 2
            WHEN name ILIKE '%TERM' THEN 1
            WHEN name ILIKE '%TERM%' THEN 0
       END search_rank
FROM TABLE_NAME -- i.e. tags, categories, etc.
WHERE slug ILIKE '%TERM%'
ORDER BY search_rank DESC
```

Reply:
I think this is a good idea, I agree with you and it would have been good to start with this, but at this point in time further optimisations and refactors for this need to be pushed to a later version, otherwise I will never get off this work :laughing: The requirements for this search changed a lot over time, hence these new additions rather than an overhaul. I also wanted to take advantage of existing code (e.g. things like `DiscourseTagging`) as much as possible for handling permissions, though doing a direct query would be faster.